Added RFC0060 - Enable creating and modifying of FilesContainers with…

SAFEShane opened the pull request:
…in the SAFE Browser environment

SAFEShane commented:
Please note that merging of this is pending discussion in the following SAFE forum topic: https://safenetforum.org/t/enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment/31088, I'll update this PR with another comment once discussion has ceased.

SAFEShane commented:
After discussion we have also decided to add a `type` field to the `files_container_add_from_raw`. This PR is now in a mergeable state and can be considered by the Maidsafe team for inclusion.

iancoleman commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 *.sublime-project
 *.sublime-workspace
 *.swp
+.idea
```
Should this (and other personal dev tools things already in .gitignore) be moved out of this repository? Typically these would belong in a `.gitignore_global` file. In git this is the configuration option called `excludesfile`.

https://intellij-support.jetbrains.com/hc/en-us/community/posts/115000154070-How-to-gitignore-idea-files?page=1#community_comment_115000258150

SAFEShane replied:
Hi Ian, I've removed this entry. You're quite right, it was just a reflex action for me to add it. I've left the others in there for now as I don't want to break anyone else's flow - it seems out of scope for this RFC, but I'd be happy to sponsor another pull request removing them.

SAFEShane commented:
@dirvine hey, it would be great to have some feedback on this, and potentially a merge if there's no more feedback left.

The actual features contained within would be trivial to implement and would unlock a lot of functionality for in-browser app developers.

Cheers

dirvine commented:
@SAFEShane  Don't know how I missed this. Let's get @joshuef  and @bochaco  to comment as well. I will ping them on slack as well.

joshuef commented on text/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment.md:
```diff
@@ -0,0 +1,89 @@
+* `files_container_add`
+* `files_container_add_from_raw`
+
+Logic would dictate that the browser would call `files_container_create` to create the files container, however first argument to this method `location: &str` requires a path to a directory or file which will be uploaded to the FilesContainer. The same issues also exist for `files_container_sync` and `files_container_add`.
```
Here I would say that we should probably just make that `location` optional. That keeps the `create` function simple, but also more true (otherwise I’d say `create` should always create empty and we should have another func for `from_path` eg. But i think we can keep this API simple and just make this optional 👍)

joshuef commented on text/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment.md:
```diff
@@ -0,0 +1,89 @@
+
+`files_container_add_from_raw` is usable, but it always defaults the `type` parameter (which is used as the HTTP `Content-Type` value in the browser to `Raw` regardless of the filetype uploaded..
+
+I propose the addition of 1 new API to be added to the NodeJS library (and thereby allowed for use via the SAFE Browser):
```
`I propose the addition of 1 new API to be added to the NodeJS library` -> `I propose the addition of 1 new API to be added to the NodeJS library, and the underlying rust SAFE-API`. I’d get that in just to clarify we want these changes throughout the api stack.

joshuef commented:
Ah, also you need to run this wee script it seems (from travis):

```
number of changes to the RFCs-by-status.md = 6
137This has failed because it appears that there has been a change to a
138RFC document status, which has not been reflected in your PR.
139Please run the "./generate_by_status.sh > RFCs-by-status.md" and commit the updates
140as part of your PR
```

SAFEShane commented:
@joshuef thanks for your comments, I've updated the RFC to reflect the request changes and updated the `RFCs-by-status.md` file. All checks are now passing.

bochaco commented on text/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment.md:
```diff
@@ -0,0 +1,76 @@
+
+Setting a type of `text/html` should result in a downloaded `Content-Type` of `text/html`.
+
+The `Content-Type` can also contain a charset value (and any number of mime related data, separated by semi-colons.). Setting a type of `text/html; charset=utf-8` should result in a downloaded `Content-Type` of `text/html; charset=utf-8`
```
We don't have support for charset or other type of mime type info in the XOR-URL, therefore I'm not sure we can support these things but just a mime type. Unless we include in the scope of this proposal how to expand XOR-URL encoding.

SAFEShane replied:
Okay, that's fine, I can amend that easily. It would have been nice to have the hinting but it's a very far away secondary concern compared to the mime type.

bochaco commented on text/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment.md:
```diff
@@ -0,0 +1,76 @@
+
+## Summary
+
+The NodeJS libraries work via the existing SAFE network CRUD conventions: we take files, we upload them to a network and generate an XORURL we can reference as a result. In the browser we do not have access to a native file sytem, so the underlying NodeJS libraries which the browser calls must allow the uploading of raw bytes, and the creation of empty FilesContainers.
```
> In the browser we do not have access to a native file sytem, so the underlying NodeJS libraries which the browser calls must allow the uploading of raw bytes

Is this the case then with new browser? I know that [works fine on an electron app](https://github.com/bochaco/electron-neon-safe/blob/master/renderer.js#L41), but haven't tried on the browser, have you @SAFEShane ?

SAFEShane replied:
@bochaco I never tested it because it didn't seem like it would work, but: oh gods:

```
v.files_container_create("testfolder/", "safe://bananatest", false, false);
VM478:1 Uncaught Error: internal error in Neon module: Failed to create FilesContainer: FileSystemError("Couldn\'t read metadata from source path (\'testfolder/\'): The system cannot find the file specified. (os error 2)")
```

```
v.files_container_create("C:\\Users\\Shane\\Desktop", "safe://bananatest", false, false);
(3) ["safe://hnyynyiqpaqxwwmq3ecceuj546hoetx4hs7ztczmshf9rz4aabww6de66ybnc", {…}, {…}]
```

```
PS C:\Users\Shane> safe cat hnyynyiqpaqxwwmq3ecceuj546hoetx4hs7ztczmshf9rz4aabww6de66ybnc
Files of FilesContainer (version 0) at "hnyynyiqpaqxwwmq3ecceuj546hoetx4hs7ztczmshf9rz4aabww6de66ybnc":
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| Name                                                                  | Size   | Created              | Modified             | Link                                                              |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| /safe:/bananatest/88230711_1826831327441917_8628637686947032041_n.jpg | 52136  | 2020-03-02T14:26:10Z | 2020-03-02T14:26:10Z | safe://hbwybon9wkw3ugjp48mmtobdn6jwfmej8bp7wato984tpfp7nrb839n5dn |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| /safe:/bananatest/SAFE Browser Alpha.lnk                              | 2532   | 2020-03-02T14:26:10Z | 2020-03-02T14:26:10Z | safe://hbyyyynsu6do1kd11xxzhmd5hzmx4957f9uadt5erysf9oarcag71x8fh6 |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| /safe:/bananatest/desktop.ini                                         | 282    | 2020-03-02T14:26:10Z | 2020-03-02T14:26:10Z | safe://hbhydynomxgm3fzy73f6qsancj1uwz39j31sd4f8c7utqwap7pyf9eye36 |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| /safe:/bananatest/screen1.png                                         | 75122  | 2020-03-02T14:26:10Z | 2020-03-02T14:26:10Z | safe://hbwynydyb11d53a6amu7hfyw89weurgmg676q3qjpn8yiwhbzfo35pz433 |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| /safe:/bananatest/screen2.png                                         | 23576  | 2020-03-02T14:26:10Z | 2020-03-02T14:26:10Z | safe://hbwynynh4mzbt1qon1qchf48waeannkja3ofij3hkfa87k7wgzjih6noc7 |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| /safe:/bananatest/screen3.png                                         | 13719  | 2020-03-02T14:26:10Z | 2020-03-02T14:26:10Z | safe://hbwynydybd5tke9pnbdbd5d64zw495uju9eaucof1sc6f6wax4qr7j3ts4 |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| /safe:/bananatest/screen4.png                                         | 124803 | 2020-03-02T14:26:10Z | 2020-03-02T14:26:10Z | safe://hbwynydy4gj9eujb6yj4tto7kyburiuct6ifczskq5d83ox3e568it888a |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
| /safe:/bananatest/screen5.png                                         | 515580 | 2020-03-02T14:26:10Z | 2020-03-02T14:26:10Z | safe://hbwynynbp87c65bffhaya1z3crgzs1k4f7gz8n8dkmty9wa4rg39h6dwwx |
+-----------------------------------------------------------------------+--------+----------------------+----------------------+-------------------------------------------------------------------+
```

This is something we should disable *immediately*. There should be no reason, no reason at all, for every single website I access in the SAFE browser to have unfettered access to my entire file system. That seems like a huge, huge, HUGE private violation.

bochaco replied:
Ok, so it works then as I suspected. 
So I guess what we need to discuss now is if we want to disable access to the file system from the browser as part of the sand-boxing, always or perhaps enable it only if the user allows it??...

SAFEShane replied:
So, I think the issue here is that we're using a single library (the node-js bindings) for both the browser and the node-js environment. Within node-js (or via an Electron app) you've already gone through a trust model: You've installed the executable and accepted the admin permissions to give it access to what Electron needs access to in order to function. With this, I've gone to a website and it has defacto access to the entire operating system's files.

Without separating out the function scopes of the browser and the underlying Electron / node-js app I'm not sure we can fix this: *the code that runs in node-js is fine, but within the browser sandbox it's not_. 

Would you be *able* to sandbox the function just when it runs in the context of the browser?

If so, I don't think there is any reason for the browser to have access to the filesystem. Uploading of individual files runs within the context of Windows Explorer windows calls, and I can't see any reason a web app would need access to the users' file systems in this way. I think a complete sandbox is necessary, lest we try to whitelist or blacklist but accidentally leave a privacy vulnerability in there - or forget about this and reintroduce it in future when editing the node-js libraries.

bochaco replied:
I think it's easy to hide this on the browser, the wrapper in the browser can be changed to hide them AFAIK

SAFEShane replied:
Okay, my concern was that the call was being proxied down to the node-js / Electron libraries which would have access to the file system as part of its functioning. If it's possible to pull the file system out of scope just for the function calls made within a web browser page, that's great.

Do you think its worth reviewing the rest of the node-js functions to see if there is anything else they bring in to scope which should be sandboxed by default?

bochaco replied:
I don't think there are any other API which read anything local, only the ones related to files and FilesContainers so far.

joshuef replied:
Yeh super easy to disable. We can just overwrite the function on init.

That'll just mean devs will have to be going through `FileReader` APIs for that.

SAFEShane commented:
Cheers for that @bochaco, @joshuef the PR is now ready for merge. :)

SAFEShane commented:
@joshuef Can I chase you guys for a specific part of this?

```
#### files_container_create

This method will have its first argument (`location: &str`) made optional, with the default being an empty files container.
```

This (incredibly minor!) feature alone would unlock a whole bunch of functionality!

bochaco commented:
I just raised this one, please take a look at it: https://github.com/maidsafe/safe-api/pull/457, once we are happy with it I can raise the PR for safe-nodejs binding.
